OpenTrons/opentrons-api
# refactor(robot-server): Remove the `ProtocolStore` pickle column from the SQL database

# Overview

This PR refactors the way that `robot-server` stores uploaded protocols across reboots.

The new way is computationally slower, but it eases development burden by minimizing the amount and complexity of stuff that we store persistently. In particular, it avoids a whole bunch of migration and versioning concerns.

Closes #9858.

# Changelog

## Production code

* Remove the `ProtocolSource` pickle column from the SQL `protocols` table.
* Make `ProtocolStore` keep `ProtocolSource`s in-memory, never committing them to any kind of persistent storage.
* When a the `ProtocolStore` is initialized for the first time, compute all necessary `ProtocolSource`s by reading the stored protocol source files.

## Integration tests

* Adjust some timeouts. These tests were timing out spuriously on my poor 2018 MacBook Air.
* When testing persistence, broaden the fields checked from just `id` and `createdAt` to everything but `analyses`. This is important to cover because of the code that I've changed in this PR. If there's a bug in this PR, the fields `files`, `protocolType`, and `metadata` could be wrong after a reboot.

# Review requests

Architectural opinions would be appreciated.

This arguably breaks our architectural convention of every unit either being a pure collaborator, or a pure worker—never a mixture of both. `ProtocolStore` remains primarily a worker, *except* for the new `.rehydrate()` method, which depends on a `ProtocolReader` interface. It depends on whether you count a `@classmethod` as part of "the unit," I guess.

I'm also leaning totally on integration tests to cover this behavior. One problem with this is that it's suffering from long test times. An alternative could have been to do quasi-integration "smoke tests" like we have for `protocol_runner`. But then, it would be unclear to me what belongs in those smoke tests, and what belongs in the integration tests.

# Manual testing

Same tests as before:

> 1. Upload some protocols with several files.
> 2. Inspect the responses from the upload, and from `GET /protocols` and `GET /protocols/{id}`. Make sure they all seem correct: no weird datetimes, no unexpected `null`s, and so on.

If you were testing on prior `edge` builds, you might need to SSH into your robot and run `rm -rf /data/opentrons_robot_server` to wipe the old database.

# Risk assessment

Medium.

I think removing the pickle column is a big improvement, but this still feels fragile. `ProtocolStore.rehydrate()` needs to know the filesystem layout of protocol source files, which is currently an implementation detail of the router. If these get out of sync, the error could cause `robot-server` to fail to fully initialize.
